design/design-generic-references.md

```diff
@@ -0,0 +1,477 @@
+
+```yaml
+apiVersion: datasync.aws.upbound.io/v1beta1
+kind: Task
+metadata:
+  name: datasync-task-example
+spec:
+  forProvider:
+    .
+    .
+    sourceLocationArnRef:
+      name: ref-source-location
+      apiVersion: apiextensions.crossplane.io/v1alpha1
+      kind: Referable
+      fieldPath: status.value
+```
+
+Of course, it should be possible to declare a source for the `Referable` 
+using labels:
+
+```yaml
+apiVersion: apiextensions.crossplane.io/v1alpha1
+kind: Referable
+metadata:
+  name: ref-source-location
+spec:
+  source:
+    matchingLabels:
+      foo: bar
+    apiVersion: datasync.aws.upbound.io/v1beta1
+    kind: LocationS3
+    fieldPath: metadata.annotations[crossplane.io/external-name]
+```
+
+### Composition
+
+Instead of declaring resources with [observe only management policy], we would
+like to enable adding values from external resources to the environments,
+similar to how environment configs are referred.
+
+```yaml
+apiVersion: apiextensions.crossplane.io/v1
+kind: Composition
+# Removed for Brevity
+spec:
+  environment:
+    environmentConfigs:
+      - type: Reference
+        ref:
+          name: source-location
+          apiVersion: datasync.aws.upbound.io/v1beta1
+          kind: LocationS3
+          fromFieldPath: metadata.annotations[crossplane.io/external-name]
+          toFieldPath: source.location # path where the value is inserted into the environment
+
+  resources:
+  # Removed for Brevity
+```
+
+or if the name of resource is unknown/irrelevant, we can use selectors:
+
+```yaml
+apiVersion: apiextensions.crossplane.io/v1
+kind: Composition
+# Removed for Brevity
+spec:
+  environment:
+    environmentConfigs:
+      - type: Selector
+        selector:
+          matchLabels:
+            foo: bar
+          apiVersion: datasync.aws.upbound.io/v1beta1
+          kind: LocationS3
+          fromFieldPath: metadata.annotations[crossplane.io/external-name]
+          toFieldPath: source.location # path where the value is inserted into the environment
+
+  resources:
+  # Removed for Brevity
+```
+
+After initializing the environment, an appropriate patch & transformation
+strategy can be applied to set managed resource fields.
+
+__**Out of proposal scope**__: renaming `environmentConfigs` field to something
+more generic, e.g. `sources`.
+
+### Advantages
+
+* All referred values are still declared on managed resources using already
+  familiar mechanism
+* Optional `Referable` instances can be watched and used by non-Crossplane controllers as
+  well
+* Avoid system pollution with [observe only management policy] objects
+* Using compositions, reference could be set even if provider does not support
+  the proposed reference mechanism
+
+### Disadvantages
+
+* Support for `Referable` type requires new controller on Crossplane side
+* Provider RBAC rules should be extended to include read permission 
+  for `Referable` type
+* Providers needs to support the new reference mechanism on their side
```

Comment:
Maybe I'm missing this: How can custom transformation/extraction logic on the data path from source to target be handled? We are preserving the `reference.ResolutionRequest.Extract`, right? This proposal is mainly for making the reference target dynamic and rest of the data path is (mostly) untouched?